### DISCUSSION TITLE
TypeScript: What type for register (instead of any) when calling it as a prop?

### DISCUSSION THREAD

**Discussion by ptmkenny** [react-hook-form/react-hook-form#4426] [Q&A] [answered]
I read the [TypeScript documentation](https://react-hook-form.com/ts), but I don't understand which type to use for `register` when using it as a prop.  This is probably a very easy question but I'm relatively new to TypeScript and would appreciate any help.

For example:

```
interface MyProps {
  register: any, // What type should I use here?
}

const UserFormHandedness: React.VFC<MyProps> = ({ register }: MyProps) => (
  <IonItem>
    <IonSelect
      ref={register}
      id="edit-field-handedness"
      name="field_handedness"
    >
      <IonSelectOption value="r">Right-handed</IonSelectOption>
      <IonSelectOption value="l">Left-handed</IonSelectOption>
    </IonSelect>
  </IonItem>
);
```

**Comment by bluebill1049** [comment]
https://react-hook-form.com/ts#UseFormMethods

`UseFormMethods['register']`

**Reply by goberlan** [reply]
You provide this link, but nowhere is `UseFormMethods['register']` located on that page? And no longer is UseFormMethods even.

**Reply by goberlan** [reply]
Would: https://react-hook-form.com/ts#UseFormProps be the new solution?

**Comment by ptmkenny** [comment]
What if I call the component like `<UserFormHandedness register={register({ required: true })} />`?

This causes an error like: 

`<html>TS2322: Type '(ref: (HTMLSelectElement &amp; HTMLTextAreaElement &amp; CustomElement&lt;FieldValues&gt; &amp; HTMLInputElement) | null) =&gt; void' is not assignable to type '{ &lt;TFieldElement extends FieldElement&lt;FieldValues&gt;&gt;(rules?: Partial&lt;{ required: string | ValidationRule&lt;boolean&gt;; min: ValidationRule&lt;string | number&gt;; ... 7 more ...; setValueAs: (value: any) =&gt; any; }&gt; | undefined): (ref: (TFieldElement &amp; Ref) | null) =&gt; void; (name: string, rules?: Partial&lt;...&gt; | undefined): void...'.<br/>Types of parameters 'ref' and 'rules' are incompatible.<br/>Type 'Partial&lt;{ required: string | ValidationRule&lt;boolean&gt;; min: ValidationRule&lt;string | number&gt;; max: ValidationRule&lt;string | number&gt;; ... 6 more ...; setValueAs: (value: any) =&gt; any; }&gt; | undefined' is not assignable to type '(HTMLSelectElement &amp; HTMLTextAreaElement &amp; CustomElement&lt;FieldValues&gt; &amp; HTMLInputElement) | null'.<br/>Type 'undefined' is not assignable to type '(HTMLSelectElement &amp; HTMLTextAreaElement &amp; CustomElement&lt;FieldValues&gt; &amp; HTMLInputElement) | null'.`

**Comment by dagstuan** [comment]
as @ptmkenny mentions. Using`UseFormMethods['register']` only works if one passes `register` directly to the child. What should be used when passing props such as `register={register({ required: true })}` @bluebill1049 ?

**Comment by ptmkenny** [comment]
After upgrading to V7, it looks like the way to do this now is:
`
  register: UseFormRegister<MyInterface>,
`

**Reply by jmayergit** [reply]
The problem with you answer is `MyInterface` should not be defined in the Input component. Imagine having to import all the interfaces into your Input component, from the consumers of the Input component who define those interfaces..

**Reply by ptmkenny** [reply]
@jmayergit I'd be happy to see a better way if you've found one.

**Reply by Kevin-Carlos** [reply]
I think it'd be the same for `control`. (note haven't tested this for register)

```typescript
type MyType<T extends FieldValues> = {
  register: UseFormRegister<T>;
}
```

which I think might work since the signature is: `export declare type UseFormRegister<TFieldValues extends FieldValues>...`

**Reply by ptmkenny** [reply]
@Kevin-Carlos Ok, based on your example, I was able to do this:

```
interface MyProps<T extends FieldValues> {
  register: UseFormRegister<T>,
}

const MyInput: React.VFC<MyProps<MyInterface>> = ({ register }: MyProps<MyInterface>) => (
```

But, previously I have been doing this:

```
interface MyProps {
  register: UseFormRegister<MyInterface>,
}

const MyInput: React.VFC<MyProps> = ({ register }: MyProps) => (
```

which seems much less verbose.

I don't understand the use of `<T>` very well, so what's the benefit of this approach?

**Reply by Kevin-Carlos** [reply]
Benefit of this approach is that you're specifying your input component generically so that the typing can be inferred:

> Note: Am using `control` here as my example since I'm trying to figure out how to generically type my components that use `control`, but the sentiment and props are _mostly_ built using the same generics under the hood in RHF.

![image](https://user-images.githubusercontent.com/26825980/141217796-da53ea46-6205-4fdb-b2be-7b5c4263fcec.png)

Notice how when I hover over control though, it has the same typings as what I declared in the `useForm`.

Same as if I actually declare the form value typings in a `type/interface`

![image](https://user-images.githubusercontent.com/26825980/141218092-157ae620-0d2c-40d8-aabd-f0339f7f1b3d.png)

**Comment by ritikbanger** [comment, marked as answer]
Hello @ptmkenny, if you are calling register as a prop in a custom component or in some custom form field in Typescript, then you can use the below code.

First of all import the UseFormRegister and FieldValues from 'react-hook-form'

`import {UseFormRegister, FieldValues } from 'react-hook-form'`

After that, define the type of register as

`register: UseFormRegister<FieldValues>`

and boom, you don't need to use 'any' type for register.

**Reply by ptmkenny** [reply]
Perfect, do you happen to know if there is a similar type for `<Control>` in the same situation?

**Reply by goberlan** [reply]
Yes I would like to know this, as well.

**Reply by goberlan** [reply]
Maybe this: https://react-hook-form.com/ts#Control

**Reply by Moshyfawn** [reply]
> Maybe this: https://react-hook-form.com/ts#Control

Is `import type { Control } from 'react-hook-form'` what you are looking for?
```
interface Props {
  control: Control<FieldValues>
}
```

**Reply by andreynav** [reply]
@ritikbanger this suggestion works only in case if you use as well: 
```
... = useForm<FieldValues>(...)
...
const onFormSubmit = (data: FieldValues) => {...}
```

But if you want to explicitly indicate a custom type [custom type](https://legacy.react-hook-form.com/api/useform) like 
```
type DialogDataT = {
	message: string
}

... = useForm<DialogDataT>(...)
```
it causes the next problems like this:

`Type 'UseFormRegister<DialogDataT>' is not assignable to type 'UseFormRegister<FieldValues>`

**Reply by killerburn69** [reply]
How to fix Type 'UseFormRegister<DialogDataT>' is not assignable to type 'UseFormRegister<FieldValues> this? can you show me the solution, please?

**Reply by andreynav** [reply]
> How to fix Type 'UseFormRegister' is not assignable to type 'UseFormRegister this? can you show me the solution, please?

So, this is my question too...

The internal type `FieldValues` does not allow you to strictly restrict the type of incoming data - https://take.ms/NhDRL 
And type `FieldValues` doesn't infer the data types automatically, so it's just the same like `data: unknown` and allow you do not see TS errors, but isn't solve the initial problem.

Or I can't found the solve of the problem...

**Reply by jordanrigo** [reply]
Have you found a solution for:
`Type 'UseFormRegister<DialogDataT>' is not assignable to type 'UseFormRegister<FieldValues>`

**Reply by andreynav** [reply]
Unfortunately no

**Reply by techdan-to** [reply]
Any luck on this?

**Reply by ritikbanger** [reply]
This is correct, I also faced such errors when the type is not explicit. You should give the type explicitly.

@bluebill1049 can you help?

**Reply by IanVS** [reply]
This also comes up when attempting to create a "partial" form component that can be reused in different other forms: https://github.com/orgs/react-hook-form/discussions/4672#discussioncomment-5982735.

**Reply by Henry-GLS-NXT** [reply]
In a quick an dirty way, it can be solved like this

` register={register as unknown as UseFormRegister<FieldValues>}` in order to avoid specific keys

**Comment by DiFuks** [comment]
Still have problem :( https://codesandbox.io/s/react-hook-form-control-forked-w39qfq?file=/src/App.tsx

**Comment by uKiJo** [comment]
i'm still getting `Type UseFormRegister<Type> is not assignable to type UseFormRegister<FieldValues>`,, any solution to that?

**Reply by omeryaz** [reply]
I have been able to solve this issue with the following setup:

Input Props:
```
export interface InputProps<T extends FieldValues>{
  name: Path<T>;
  register: UseFormRegister<T>;
// ...rest of the types
}
```

Input Component:
```

const Input = <T extends FieldValues>({
  name,
  register,
// ...rest of the props
}: InputProps<T>) => (
    <InputWrapper >

      <StyledInput
        {...register(name)}
      />

    </InputWrapper>
);


export default Input;
```

Form Setup:
```
const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<SignupFormValues>({
    defaultValues: {
      email: '',
      password: '',
      phoneNumber: '',
    },
    resolver: yupResolver(validationSchema),
  });
```

Input Use:
```
<Input
   name="email"
   register={register}
 />
```
It works with both of my forms:

<img width="654" alt="Screenshot 2024-01-18 at 14 14 32" src="https://github.com/react-hook-form/react-hook-form/assets/103966057/81a944f0-6453-48dd-bb07-adf3b1656e64">
<img width="649" alt="Screenshot 2024-01-18 at 14 14 51" src="https://github.com/react-hook-form/react-hook-form/assets/103966057/45ae4f4e-0b56-407a-a238-be06dd03fdd0">

Version 7.49.2 of react-hook-form
